contracts/contracts/MessageProcessor.sol

```diff
@@ -82,119 +79,92 @@ contract MessageProcessor is Ownable, SnarkCommon, Hasher, CommonUtilities, IMes
       revert NoMoreMessages();
     }
 
-    // The state AccQueue must be merged
-    if (!poll.stateMerged()) {
-      revert StateNotMerged();
-    }
-
     // Retrieve stored vals
-    (, uint8 messageTreeSubDepth, uint8 messageTreeDepth, uint8 voteOptionTreeDepth) = poll.treeDepths();
+    (, uint8 voteOptionTreeDepth) = poll.treeDepths();
     // calculate the message batch size from the message tree subdepth
-    uint256 messageBatchSize = TREE_ARITY ** messageTreeSubDepth;
-
-    (, AccQueue messageAq) = poll.extContracts();
-
-    // Require that the message queue has been merged
-    uint256 messageRoot = messageAq.getMainRoot(messageTreeDepth);
-    if (messageRoot == 0) {
-      revert MessageAqNotMerged();
-    }
+    uint8 messageBatchSize = poll.batchSizes();
 
+    uint256[] memory batchHashes;
     // Copy the state and ballot commitment and set the batch index if this
     // is the first batch to process
     if (numBatchesProcessed == 0) {
       uint256 currentSbCommitment = poll.currentSbCommitment();
       sbCommitment = currentSbCommitment;
-      (, uint256 numMessages) = poll.numSignUpsAndMessages();
-      uint256 r = numMessages % messageBatchSize;
 
-      currentMessageBatchIndex = numMessages;
+      poll.padLastBatch();
+      batchHashes = poll.getBatchHashes();
+      currentBatchIndex = batchHashes.length;
 
-      if (currentMessageBatchIndex > 0) {
-        if (r == 0) {
-          currentMessageBatchIndex -= messageBatchSize;
-        } else {
-          currentMessageBatchIndex -= r;
-        }
+      if (currentBatchIndex > 0) {
+        currentBatchIndex -= 1;
       }
+    } else {
+      batchHashes = poll.getBatchHashes();
```

Comment:
this is not efficient, we should be pulling the batch hashes once, not for every batch processed